presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java

```diff
@@ -595,11 +621,98 @@ public void load(LazyBlock lazyBlock)
+        {
+            this.block = requireNonNull(block, "block is null");
+            this.rowsToKeep = requireNonNull(rowsToKeep, "rowsToKeep is null");
+        }
+
+        @Override
+        public void load(LazyBlock lazyBlock)
+        {
+            if (block == null) {
+                return;
+            }
+
+            if (block instanceof LazyBlock) {
+                block = ((LazyBlock) block).getBlock();
+            }
+            lazyBlock.setBlock(filterBlock(block, rowsToKeep));
 
             // clear reference to loader to free resources, since load was successful
             block = null;
         }
+
+        private static Block filterBlock(Block block, IntArrayList rowsToKeep)
+        {
+            if (block instanceof RunLengthEncodedBlock) {
+                return block.getRegion(0, rowsToKeep.size());
+            }
+            if (block instanceof DictionaryBlock) {
+                return block.getPositions(rowsToKeep.elements(), 0, rowsToKeep.size());
+            }
+            return new DictionaryBlock(rowsToKeep.size(), block, rowsToKeep.elements());
+        }
+    }
+
+    private static Page extractColumns(Page page, int[] columns)
+    {
+        Block[] blocks = new Block[columns.length];
+        for (int i = 0; i < columns.length; i++) {
+            int dataColumn = columns[i];
+            blocks[i] = page.getBlock(dataColumn);
+        }
+        return new Page(page.getPositionCount(), blocks);
+    }
+
+    public static class BucketAdapter
+    {
+        public final int[] bucketColumns;
+        public final int bucketToKeep;
+        public final int tableBucketCount;
+        public final int partitionBucketCount; // for sanity check only
+        private final List<TypeInfo> typeInfoList;
+
+        public BucketAdapter(BucketAdaptation bucketAdaptation)
+        {
+            this.bucketColumns = bucketAdaptation.getBucketColumnIndices();
+            this.bucketToKeep = bucketAdaptation.getBucketToKeep();
+            this.typeInfoList = bucketAdaptation.getBucketColumnHiveTypes().stream()
+                    .map(HiveType::getTypeInfo)
+                    .collect(toImmutableList());
+            this.tableBucketCount = bucketAdaptation.getTableBucketCount();
+            this.partitionBucketCount = bucketAdaptation.getPartitionBucketCount();
+        }
+
+        public IntArrayList computeEligibleRowIds(Page page)
+        {
+            IntArrayList ids = new IntArrayList(page.getPositionCount());
+            Page bucketColumnsPage = extractColumns(page, bucketColumns);
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                int bucket = HiveBucketing.getHiveBucket(tableBucketCount, typeInfoList, bucketColumnsPage, position);
+                if ((bucket - bucketToKeep) % partitionBucketCount != 0) {
+                    throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format(
```

Comment:
Maybe add an custom exception class (e.g., `InvalidBucketFilesException`) and use it in both places.